### ISSUE TITLE
Do we need a reserved-memory node for the shared memory in the DTB file?

### ISSUE THREAD

**Issue by xxxken** [OP-TEE/optee_os#2576]
When I run optee, I very occasionally run into a crash "usercopy: kernel memory exposure attempt detected from ffff800007e01000 (kmalloc-128) (8192 bytes)"

The address 0xffff800007e01000 belongs to the shared memory. I am guessing the dtb file I use to boot the Android kernel does not explicitly include any information about the shared memory so I cannot see any node that reserves it under /proc/device-tree and therefore something else touched that area. But I'm not sure if this is true. Actually, I think```ptee->shm_vaddr = ioremap_cache(ptee->shm_paddr, shm_size);``` in the tee_tz_drv.c is enough.
Am I correct? What can be the possible cause?

**Comment by jenswi-linaro** [comment]
Yes, the memory should be reserved or there will be problems like what you describe.

**Comment by mysticTot** [comment]
@jenswi-linaro, @jbech-optee-adm @MrVan 

I'm trying to reserve some DRAM for optee in u-boot using 
```bash
diff --git a/board/engicam/common/board.c b/board/engicam/common/board.c
index 5dccb17cb271..533312e8fc44 100644
--- a/board/engicam/common/board.c
+++ b/board/engicam/common/board.c
@@ -13,6 +13,8 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#define CONFIG_TEE_RAM_SIZE 0x01e00000
+
 #ifdef CONFIG_ENV_IS_IN_MMC
 static void mmc_late_init(void)
 {
@@ -109,7 +111,7 @@ int board_init(void)
 
 int dram_init(void)
 {
-       gd->ram_size = imx_ddr_size();
+       gd->ram_size = imx_ddr_size() - CONFIG_TEE_RAM_SIZE;
 
        return 0;
 }
```` 
After that i boot the optee and linux .

But when I check  `cat /proc/iomem` in linux, I see it is showing `10000000-2e1fffff : System RAM`
Shouldn't it show `10000000-2C3FFFFF`, ie, `2e1FFFFF minus 01e00000`(which is the size of TZ RAM reserved in u-boot).

My board's DDR start address is 10000000 and its size is 2E200000

According to this cuurent issue [this](https://github.com/OP-TEE/optee_os/issues/1034) issue, we need to include a device tree node for optee and include `no-map` property.

I think  u-boot is not able to reserve the memory for optee.

But i see many developers  has done such reservation in u-boot, for example warp7.
[board/warp7/warp7.c +39](http://git.denx.de/?p=u-boot.git;a=blob;f=board/warp7/warp7.c;h=6ebeb08e33144a4a606a0d2f9ee0ef2811bf1c9f;hb=HEAD#l39)

Its nowhere clearly mentioned that how to reserve the memory for optee.

I'm thinking if this is not working then any new developer (like me) can assume that u-boot has reserved the RAM for optee which actually hasn't. 

Please note that I'm able to run xtest and examples with this setup. 

Please correct me if my understanding is wrong.

**Comment by MrVan** [comment]
@mysticTot `+       gd->ram_size = imx_ddr_size() - CONFIG_TEE_RAM_SIZE;` imx_ddr_size will reflect the real hardware dram size. If you reserve optee on high end of the DRAM, decrease the size with CONFIG_TEE_RAM_SIZE is ok.
uboot will use gd->ram_size to update device tree before booting linux kernel, you will not see the higher memory in `/proc/iomem`

**Comment by mysticTot** [comment]
@MrVan ,
Thanks for your reply.
We can protect dram using TZASC but Why are we  not using tzc-380 driver for protection of DRAM?
I see it is added in optee but not configured for imx boards.

**Comment by mysticTot** [comment]
@MrVan ,
Thanks for your reply.
We can protect dram using TZASC but Why are we  not using tzc-380 driver for protection of DRAM?
I see it is added in optee but not configured for imx boards.

**Comment by Emantor** [comment]
@mysticTot Currently upstream OP-TEE does not setup the TZASC(s). The downstream [imx-optee-os](https://source.codeaurora.org/external/imx/imx-optee-os/) implements support but has not been upstreamed.

**Comment by MrVan** [comment]
@MrVan @Emantor that has not been upstreamed, but major issue is need to find a good method to runtime configure tzasc without hardcoding tzasc configuration for each board.

**Comment by PratikGajera** [comment]
Hi MrVan,
Can you provide me full detailed how to configure tzasc at runtime for imx6ull board.

**Comment by Emantor** [comment]
@PratikGajera https://github.com/OP-TEE/optee_os/pull/3017

**Comment by PratikGajera** [comment]
Thanks @Emantor.
Issue(#3047) has been resolved by applying this (#3017) solution.
